src/main/java/com/example/demo/matcher/services/OrderService.java

```diff
@@ -1,23 +1,101 @@
 package com.example.demo.matcher.services;
 
 import com.example.demo.matcher.models.Order;
+import com.example.demo.matcher.models.OrderAction;
+import com.example.demo.matcher.models.OrderbookItem;
 
 import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Objects;
+import java.util.stream.Collectors;
 
 public class OrderService {
-    private ArrayList<Order> orders;
+    private final ArrayList<Order> orders;
+
+    public OrderService() {
+        orders = new ArrayList<>();
+    }
 
     public ArrayList<Order> get() {
         return orders;
     }
 
     public void add(Order order) {
-
+        orders.add(order);
     }
 
     public void remove(Order order) {
+        orders.remove(order);
+    }
 
+    public ArrayList<OrderbookItem> getOrderbook(OrderAction action, String accountId) {
+        // filter the order list by action
+        ArrayList<Order> filtered = this.get().stream().filter(order -> order.getAction() == action && Objects.equals(order.getAccountId(), accountId)).collect(Collectors.toCollection(ArrayList::new));
+        return makeOrderbook(filtered, action);
     }
 
+    public ArrayList<OrderbookItem> getOrderbook(OrderAction action) {
```

Comment:
You should "program to the interface", so return `List` instead of `ArrayList` (same for variable assignments), and only specify the type when you create an object ie `new WhateverList();`

Reply:
Okay dokey :) I guess that means that we can be more generic, and change which list implementation we're using down the line without having to refactor everything.